src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -888,12 +983,48 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
 
                         if (status.ErrorCode == SecurityStatusPalErrorCode.ContextExpired)
                         {
-                            return 0;
+                            receivedEOF = true;
+                            break;
                         }
 
                         throw new IOException(SR.net_io_decrypt, SslStreamPal.GetException(status));
                     }
+
+                    if (decryptedCount > 0)
+                    {
+                        // This will either copy data from rented buffer or adjust final buffer as needed.
+                        // In both cases _decryptedBytesOffset and _decryptedBytesCount will be updated as needed.
+                        processedLength += CopyDecryptedData(buffer);
+                        if (_decryptedBytesCount > 0)
+                        {
+                            // We have more decrypted data after we filled provided buffer.
+                            break;
+                        }
+
+                        buffer = buffer.Slice(decryptedCount);
+                    }
+
+                    if (processedLength == 0)
+                    {
+                        // We did not get any real data so far.
+                        continue;
+                    }
+
+                    if (!HaveFullTlsFrame(out payloadBytes))
+                    {
+                        // We don't have anuther frame to process but we have some data to return to caller.
+                        break;
+                    }
+
+                    TlsFrameHelper.TryGetFrameHeader(_internalBuffer.AsSpan(_internalOffset), ref _lastFrame.Header);
+                    if (_lastFrame.Header.Type != TlsContentType.AppData)
+                    {
+                        // Alerts, handshake and anything else will be processed separately.
```

Comment:
Why? Seems like it's better to process these sooner rather than later. And there could be more data in subsequent frames as well, right?

Reply:
I'm worried about compatibility. Goal of this change is efficiency, not behavior changes. While I don't have specific example (besides the AUX tests) currently all the additional service frames would be processed after handing out data received prior.

Reply:
Ok, seems reasonable. Can we at least add a comment that this probably isn't strictly necessary?